### PR TITLE
Periodic check: Ensure a PR isn't stuck 

### PR DESCRIPTION
**Pull request by jameesjohn**
## Explanation

This PR ensures that a PR isn't stuck without having any assigned reviewers. it is one of the steps in the process of ensuring that a PR doesn't get stuck.
## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by jameesjohn** [comment]
PTAL @ankita240796 
Thanks.

**Comment by jameesjohn** [comment]
This feature has been tested here: https://github.com/jameesjohn/comment-on-pr/pull/8

File: .eslintrc [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "eslint:recommended",
   ],
   "parserOptions": {
-    "ecmaVersion": 8,
+    "ecmaVersion": 9,
```
Comment (ankita240796) [inline .eslintrc:6]: Can you please explain why is this required?

File: index.js [hunk 2]
```diff
@@ -54,4 +55,12 @@ module.exports = (robot) => {
       await checkMergeConflictsModule.checkMergeConflictsInAllPullRequests(context);
     }
   });
+
+
+  robot.on('schedule.repository', async context => {
+    if (whitelistedAccounts.includes(context.repo().owner.toLowerCase())) {
+      console.log('Periodic Check...');
```
Comment (ankita240796) [inline index.js:62]: Add a log which is more explanatory -- i.e explain what the periodic check is exactly doing?

File: lib/periodicChecks.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      { ...prParams, assignees: unAssignedReviewers });
+
+    // Ping newly assigned reviewers.
+    await context.github.issues.createComment(
+      { ...prParams, body: botComment });
+  }
+};
+
+module.exports.assignReviewers = async (context) => {
+  // By default, this returns only open PRs
```
Comment (ankita240796) [inline lib/periodicChecks.js:29]: Add a full stop in the comment end.

File: lib/periodicChecks.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  const pullRequestsPromiseObj = await context.github.pullRequests.getAll(
+    context.repo({ per_page: 60 })
```
Comment (ankita240796) [inline lib/periodicChecks.js:31]: Does this mean we will get a paginated response? How is the number 60 chosen?

File: lib/periodicChecks.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+const assignUnAssignedReviewers = async (pullRequest, context) => {
```
Comment (ankita240796) [inline lib/periodicChecks.js:1]: `assignUnAssignedReviewers` -> `assignUnassignedReviewers`

File: lib/periodicChecks.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  let botComment = 'PTAL ';
+  const prParams = context.repo(
+    {
+      number: pullRequest.number,
+    }
+  );
+
+  const unAssignedReviewers = reviewers.filter(reviewer => {
+    if (!assignees.includes(reviewer)) {
+      botComment += `@${reviewer} `;
```
Comment (ankita240796) [inline lib/periodicChecks.js:14]: Can we add make the comment more descriptive - something along the lines: Hey @x, @y, @z, this PR is waiting on your review, can you PTAL? Thanks!

Also, one thing which is missing is - if the PR already has changes requested tag on it, it would be better to assign the author rather than the reviewer.

**Comment by ankita240796** [comment]
Hey @bansalnitish @DubeySandeep @kevinlee12, it would be great if you all could take a pass as well, Thanks!

File: lib/periodicChecks.js (hunk 3 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Done

File: lib/periodicChecks.js (hunk 4 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:37]: Yes, we will get a paginated response.
I chose 60 because that is what is used in a previous implementation so I decided to follow suit.

File: lib/periodicChecks.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
+const assignUnassignedReviewers = async (pullRequest, context) => {
+  const assignees = pullRequest.assignees.map(user => user.login);
+  const reviewers = pullRequest.requested_reviewers.map(user => user.login);
+
+  // If the PR author is assigned, then don't assign other reviewers.
+  if (assignees.includes(pullRequest.user.login)) {
+    return;
+  }
+
+  let botComment = 'Hey ';
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:10]: Can we use a string template here (or in any such places), as it will make it easier for us to understand the whole message? (not sure, @ankita240796 any thoughts?)

File: lib/periodicChecks.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  const prParams = context.repo(
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:11]: use full form (As used in other places) pullRequestParams

File: lib/periodicChecks.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    {
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:12]: minor nit/suggestion: to me this looks more simpler:
```
context.repo({
    number: ...,
});

```

File: lib/periodicChecks.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      number: pullRequest.number,
+    }
+  );
+
+  const unassignedReviewers = reviewers.filter(reviewer => {
+    if (!assignees.includes(reviewer)) {
+      botComment += `@${reviewer}, `;
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:19]: This should not be a part of the filter function, can we generate it separately? (for readability)

File: lib/periodicChecks.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      return true;
+    }
+  });
+  if (unassignedReviewers.length > 0) {
+    await context.github.issues.addAssigneesToIssue(
+      { ...prParams, assignees: unassignedReviewers });
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:25]: What if the unassigned reviewer has already approved the PR? will it avoid adding those reviewers?

File: lib/periodicChecks.js (hunk 7 above)
Comment (DubeySandeep) [inline lib/periodicChecks.js:10]: I would suggest having `Hi` instead of `Hey`. (`Hi` is more formal)

File: lib/periodicChecks.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+module.exports.assignReviewers = async (context) => {
+  // By default, this returns only open PRs.
+  const pullRequestsPromiseObj = await context.github.pullRequests.getAll(
+    context.repo({ per_page: 60 })
+  );
+
+  const arrayOfOpenPullRequests = pullRequestsPromiseObj.data;
+
+  arrayOfOpenPullRequests.forEach(async (pullRequest) => {
+    const pullRequestDetailsPromise = await context.github.pullRequests.get(
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:43]: this will make https call one at a time? if yes, then do we have any API to get context for multiple PRs at once?

File: lib/periodicChecks.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      context.repo({ number: pullRequest.number })
+    );
+    const pullRequestDetails = pullRequestDetailsPromise.data;
+    await assignUnassignedReviewers(pullRequestDetails, context);
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:47]: ditto here can we assign reviewers to a list of PRs at once? (not sure, can you please check whether that's possible?)

File: lib/periodicChecks.js (hunk 11 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: If the reviewer has already approved the PR, they wouldn't get listed in `pullRequest.requested_reviewers`

File: lib/periodicChecks.js (hunk 12 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Hmm...
So I just checked this and it turns out the previous call returned all the information that would be required, hence the new call is redundant.
Thanks.

File: lib/periodicChecks.js (hunk 13 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: The reviewers may not be the same so I don't think this will be feasible

File: lib/periodicChecks.js (hunk 8 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Done

File: lib/periodicChecks.js (hunk 7 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Using a template string here will cause linting errors since there is no javascript interpolation.

**Comment by jameesjohn** [comment]
PTAL @DubeySandeep @ankita240796 I've addressed all comments.

File: index.js [hunk 14]
```diff
@@ -3,14 +3,15 @@ const createScheduler = require('probot-scheduler');
...
+const periodicChecksModule = require('./lib/periodicChecks');
 const whitelistedAccounts = (
   (process.env.WHITELISTED_ACCOUNTS || '').toLowerCase().split(','));
 var pullRequestAuthor;
 
 module.exports = (robot) => {
   scheduler = createScheduler(robot, {
     delay: !process.env.DISABLE_DELAY,
-    interval: 60 * 60 * 1000 // 1 hour
+    interval: 60 * 60 * 24 * 1000 // 24 hours
```
[the remark this answers is not in the document]
Comment (DubeySandeep) [reply index.js:14]: The oppiabot-stale also works in the schedule, right? If yes, then what's the timing for it? (wanted to check how it's connected with this repo.)

File: lib/periodicChecks.js (hunk 7 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js]: I see, thanks for check. Can we generate this string in one place (like [here](https://github.com/oppia/oppiabot/blob/master/lib/checkPullRequestLabels.js#L43)) that will make it more readable, maybe?

File: lib/periodicChecks.js (hunk 13 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js]: Yeah, we can generate a comment message for all the pull request one at a time and make one call to publish all the comments (if that's possible?)?

File: lib/periodicChecks.js (hunk 11 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js]: Can you please add this info as a comment around the line where you are using `requested_reviewers`?

File: index.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (jameesjohn) [reply index.js:14]: The probot-stale works independently of the scheduler. It uses a `.github/stale.yml` file which is in the oppia project to function. It is just installed here as an app. https://probot.github.io/apps/stale/

File: lib/periodicChecks.js (hunk 13 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: I think the comment should be specific (including the reviewers). Also, I confirmed that [creating a comment](https://octokit.github.io/rest.js/v17#issues-create-comment) requires the issue or PR number, and not an array of numbers so it wouldn't be possible to publish it across multiple PRs.

File: index.js (hunk 2 above)
Comment (ankita240796) [reply index.js:62]: Please add this as a comment then.

File: lib/periodicChecks.js [hunk 15]
```diff
@@ -0,0 +1,45 @@
+const assignUnassignedReviewers = async (pullRequest, context) => {
+  const assignees = pullRequest.assignees.map(user => user.login);
+  const reviewers = pullRequest.requested_reviewers.map(user => user.login);
+
+  // If the PR author is assigned, then don't assign other reviewers.
+  if (assignees.includes(pullRequest.user.login)) {
```
Comment (ankita240796) [inline lib/periodicChecks.js:6]: What if any reviewer has requested changes on the PR but missed assigning the PR author?

File: lib/periodicChecks.js (hunk 4 above)
Comment (ankita240796) [reply lib/periodicChecks.js:37]: Which previous implementation are you referring to?

File: lib/periodicChecks.js (hunk 13 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js]: I mean is there any API to publish multiple comments at once:
```
comments: [{
  number: ...,
  body: botComment1
}, {
  number: ...,
  body: botComment1
},..]
```

This way we won't have to make multiple https calls.

Also, I to had this concern (image below): 
image
Considering this can happen, should we log exception for [403 returns](https://developer.github.com/v3/#abuse-rate-limits)? Also, maybe follow [these guidelines](https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits)?

[In general, how do we log issues (400* errors) raised while calling API?]

Comment (DubeySandeep) [reply lib/periodicChecks.js]: I think we should handle 400 or 500 errors otherwise we will be making unwanted/unauthorized calls multiple time.
`Intentionally ignoring repeated validation errors may result in the suspension of your app for abuse.` ([Source](https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-api-errors))

I'm not sure how probot handles these errors, so I just wanted to check we are aware of this.

File: index.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (DubeySandeep) [reply index.js:14]: Thanks for the explanation, @jameesjohn!

File: lib/periodicChecks.js [hunk 16]
```diff
@@ -0,0 +1,45 @@
+const assignUnassignedReviewers = async (pullRequest, context) => {
```
Comment (DubeySandeep) [inline lib/periodicChecks.js:1]: Do we need the LICENCE comment here?
/cc @seanlip

File: lib/periodicChecks.js (hunk 15 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js:6]: (Just a thought, maybe we can handle this without scheduler i.e, once a reviewer submits PR review with "request-changes" then the bot will unassign the reviewer and assign the author.)

File: lib/periodicChecks.js (hunk 16 above)
Comment (DubeySandeep) [inline lib/periodicChecks.js:1]: Do we need a unit test for these services?

File: index.js (hunk 2 above)
Comment (jameesjohn) [reply index.js]: Done.

File: lib/periodicChecks.js (hunk 15 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:6]: There is a feature in the bot sheet that assigns the user automatically when a reviewer request changes on a PR.
We can't check from the PR that a reviewer requested changes at this point (from the PR data).

File: lib/periodicChecks.js (hunk 4 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:37]: https://github.com/oppia/oppiabot/blob/6d723a95549353bb1641325f62efc95c8f420f64/lib/checkMergeConflicts.js#L76

File: lib/periodicChecks.js (hunk 15 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:6]: > (Just a thought, maybe we can handle this without scheduler i.e, once a reviewer submits PR review with "request-changes" then the bot will unassign the reviewer and assign the author.)

Yes @DubeySandeep, that is part of the oppiabot features.

File: lib/periodicChecks.js (hunk 16 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:1]: I didn't find this in other files so I'm not sure it's needed.

File: lib/periodicChecks.js (hunk 6 above)
Comment (ankita240796) [reply lib/periodicChecks.js]: Ok, so I would suggest that we should have that check in place too - i.e. assign the PR author when change is requested - since this functionality is somewhat incomplete without it. We will need to upgrade it later but it is fine to have it now since it is an essential component for this check to be complete.

File: lib/periodicChecks.js (hunk 15 above)
Comment (ankita240796) [reply lib/periodicChecks.js:6]: Yes, let us have that check in this PR as well (we will need an upgrade later but it won't be difficult to upgrade - on the other hand not having this feature would definitely be a problem, since the periodic check is somewhat incomplete without it).

Comment (jameesjohn) [reply lib/periodicChecks.js:6]: Done.

File: lib/periodicChecks.js (hunk 6 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Done. PTAL

File: lib/periodicChecks.js (hunk 16 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js:1]: I had a discussion on this with Sean and he suggested to add the LICENCE in all such files.

Comment (jameesjohn) [reply lib/periodicChecks.js:1]: Done. PTAL @DubeySandeep

Comment (jameesjohn) [reply lib/periodicChecks.js:1]: This should be to all files in this repo right?
Shouldn't that be done in a different PR?

**Comment by jameesjohn** [comment]
PTAL @ankita240796 @DubeySandeep. Thanks

File: lib/periodicChecks.js (hunk 16 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js:1]: Let's add it for this file and create an issue (and create a PR whenever possible) to fix this in other files.

**Comment by DubeySandeep** [comment]
@jameesjohn, Can you please reply to [this](https://github.com/oppia/oppiabot/pull/54#discussion_r407259892) thread?

File: lib/periodicChecks.js (hunk 13 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Ohh
I'm sorry I missed this.
> I mean is there any API to publish multiple comments at once:
No, the API does not support creating multiple comments at [once ](https://octokit.github.io/rest.js/v17#issues-create-comment)

> I'm not sure how probot handles these errors, so I just wanted to check we are aware of this.

We are not making the requests directly since we are using the [github octokit](http://github.com/octokit/rest.js/) library. And the library specifies what contents should be sent so I don't think we will fall into this issue. 
I dug deeper, and probot provides a SENTRY_DSN env variable for error logging to sentry.io https://probot.github.io/docs/configuration/

File: lib/periodicChecks.js (hunk 7 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: But this would not work for arrays since we could be pinging multiple reviewers at the same time.

File: lib/periodicChecks.js (hunk 11 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Done

File: lib/periodicChecks.js (hunk 16 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:15]: Done.
The issue can be found at https://github.com/oppia/oppiabot/issues/56

**Comment by jameesjohn** [comment]
Done @DubeySandeep.
PTAL

File: lib/handlePullRequestReview.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+// limitations under the License.
+
+const CHANGES_REQUESTED = 'changes_requested';
+
+module.exports.pullRequestReviewed = async (context) => {
+  const review = context.payload.review;
+  const prAuthor = context.payload.pull_request.user;
+
+  if (review.state === CHANGES_REQUESTED) {
+    // assign PR author to PR
```
Comment (ankita240796) [inline lib/handlePullRequestReview.js:22]: Capitalise first letter, add a fullstop in the end.

File: lib/periodicChecks.js (hunk 4 above)
Comment (ankita240796) [reply lib/periodicChecks.js:54]: Ok!

File: lib/periodicChecks.js (hunk 13 above)
Comment (DubeySandeep) [reply lib/periodicChecks.js]: > And the library specifies what contents should be sent so I don't think we will fall into this issue.

But the library itself says that don't make many comments quickly and redirects to a URL which has these guidelines ([ref](https://octokit.github.io/rest.js/v17#issues-create-comment))

> I dug deeper, and probot provides a SENTRY_DSN env variable for error logging to sentry.io https://probot.github.io/docs/configuration/

But for that we need to have a Sentry account, right?  Can you please check how to handle these http errors on probot API. Like what will happen if the add comment API doesn't work correctly.

**Comment by ankita240796** [comment]
Closing this, will be focusing on this in milestone 3.
